ensime_shared/ensime.py

```diff
@@ -136,16 +136,23 @@ def queue_poll(self, sleep_t=0.5):
 
         Value of sleep is low to improve responsiveness.
         """
+        connection_alive = True
         while self.running:
             if self.ws:
                 def logger_and_close(m):
                     self.log("Websocket exception: {}".format(m))
-                    self.teardown()
+                    if not self.running:
```

Comment:
I'm still pretty curious about this. I ran a quick test where I load up a file, let the server boot, kill the server, then issue an command. Here are the logs for current mater & your branch:

Master
```
2016-03-20 16:44:25.443928: Websocket exception: Connection is already closed.
2016-03-20 16:44:25.444037: teardown: in
2016-03-20 16:44:30.522621: type: in
2016-03-20 16:44:30.523756: send_at_position: in
2016-03-20 16:44:30.524753: path: in
2016-03-20 16:44:30.525031: 56 11
2016-03-20 16:44:30.525443: 1944
2016-03-20 16:44:30.525521: send_request: in
2016-03-20 16:44:30.525715: send: in
2016-03-20 16:44:30.525868: send: {"callId": 2, "req": {"range": {"to": 1947, "from": 1944}, "typehint": "TypeAtPointReq", "file"
: "/Users/ccoffey/workspace/open_source/ensime-server/.ensime_cache/dep-src/source-jars/org/apache/lucene/document/TextField.java
"}}
2016-03-20 16:44:30.526153: send error: socket is already closed., reconnecting...
```

remove-cycle-teardown
```
2016-03-20 17:01:36.672266: Websocket exception: Connection is already closed.
2016-03-20 17:01:36.672395: teardown: in
2016-03-20 17:01:57.979673: open_declaration: in
2016-03-20 17:01:57.980702: 74 45
2016-03-20 17:01:57.981463: 3175
2016-03-20 17:01:57.981560: path: in
2016-03-20 17:01:57.982108: send_request: in
2016-03-20 17:01:57.982248: send: in
2016-03-20 17:01:57.982375: send: {"callId": 3, "req": {"typehint": "SymbolAtPointReq", "file": "/Users/ccoffey/workspace/open_source/ensime-server/core/src/main/scala/org/ensime/indexer/IndexService.scala", "point": 3176}}
2016-03-20 17:01:57.982701: send error: socket is already closed., reconnecting...
```

In both cases it looks like the teardown code fires and jumps out of the loop on the next cycle since we only see the `Websocket exception` message once in each branch, so I went a little further and added a log line on line `144`, right above `connection_alive = False`. I never saw that log line get hit. 

Is killing the server too extreme of a test in this case, or is there something else I'm missing?